### DISCUSSION TITLE
How to get the graph or basic blocks of "Function Graph"?

### DISCUSSION THREAD

**Discussion by shinibufa** [NationalSecurityAgency/ghidra#5808] [General]
I have noticed that the number and starting addresses of basic blocks returned by ghidra.program.model.pcode.PcodeSyntaxTree.getBasicBlocks() sometimes differ from those displayed in the "Display Function Graph" window. I believe that the basic blocks shown in the "Display Function Graph" are more accurate. Is there an API available that can help me retrieve these basic blocks or the graph representation of the function?

**Comment by dragonmacher** [comment]
The Function Graph creates vertices in the `FunctionGraphFactory` using a `BasicBlockModel`:

```
private static BidiMap<CodeBlock, FGVertex> createVertices(Function function,
			final FGController controller, TaskMonitor monitor) throws CancelledException {

		BidiMap<CodeBlock, FGVertex> vertices = new DualHashBidiMap<>();
		CodeBlockModel blockModel = new BasicBlockModel(controller.getProgram());

		AddressSetView addresses = function.getBody();
		CodeBlockIterator iterator = blockModel.getCodeBlocksContaining(addresses, monitor);
		monitor.initialize(addresses.getNumAddresses());

		for (; iterator.hasNext();) {
			CodeBlock codeBlock = iterator.next();

			FlowType flowType = codeBlock.getFlowType();
			boolean isEntry = isEntry(codeBlock);
			Address cbStart = codeBlock.getFirstStartAddress();
			if (cbStart.equals(function.getEntryPoint())) {
				isEntry = true;
			}

			FGVertex vertex =
				new ListingFunctionGraphVertex(controller, codeBlock, flowType, isEntry);
			vertices.put(codeBlock, vertex);

			long blockAddressCount = codeBlock.getNumAddresses();
			long currentProgress = monitor.getProgress();
			monitor.setProgress(currentProgress + blockAddressCount);
		}

		return vertices;
	}
```

There are various block model types in Ghidra that all extends `CodeBlockModel`.   

There is no service to use the code listed above, but you can just copy and manipulate that code as needed to create code blocks using the block model in the same way.   You don't need any of the actual graph code such as the `ListingFunctionGraphVertex`, as that is specific to create a UI graph.

**Reply by shinibufa** [reply]
Thank you very much!
